sos/plugins/kubernetes.py

```diff
@@ -15,40 +15,130 @@
+
+            resources = [
+                'events', 'limitrange', 'pods', 'pvc', 'rc',
+                'resourcequota', 'services'
+            ]
+
+            for res in resources:
+                self.add_cmd_output(
+                    "{} {} {}".format(
+                        kube_bin,
+                        kube_get,
+                        res
+                    )
+                )
+            resources.remove('events')
+            # nodes and pvs are not namespaced, must pull separately
+            self.add_cmd_output([
+                "{} get -o json nodes".format(kube_bin),
+                "{} get -o json pv".format(kube_bin)
+                ]
+            )
+
+            if self.get_option('describe'):
+                resources = resources + ['nodes', 'namespaces']
+                for res in resources:
+                    r = self.get_command_output(
+                        "{} get {}".format(
+                            kube_bin,
+                            res
+                        )
+                    )
+                    if r['status'] == 0:
+                        k_list = [k.split()[0] for k in
+                                  r['output'].splitlines()[1:]
+                                  ]
+                        for k in k_list:
+                            self.add_cmd_output(
+                                "{} describe {} {}".format(
+                                    kube_bin,
+                                    res,
+                                    k
+                                )
+                            )
+
+            if self.get_option('podlogs'):
+                r = self.get_command_output(
+                    "{} get pods".format(kube_bin)
+                )
+                if r['status'] == 0:
+                    pods = [p.split()[0] for p in
+                            r['output'].splitlines()[1:]
+                            ]
+                    for pod in pods:
+                        self.add_cmd_output(
+                            "{} logs {}".format(kube_bin, pod)
+                        )
+
+    def postproc(self):
+        # First, clear sensitive data from the json output collected.
+        # This will mask values when the "name" looks susceptible of
+        # values worth obfuscating, i.e. if the name contains strings
+        # like "pass", "pwd", "key" or "token"
+        env_regexp = r'(?P<var>{\s*"name":\s*[^,]*' \
+                     r'(pass|pwd|key|token|cred)[^,]*,\s*"value":)[^}]*'
+        self.do_cmd_output_sub('kubectl', env_regexp,
+                               r'\g<var> "********"', ignore_case=True)
 
+        # Next, we need to handle the private keys and certs in describe
+        # output that is not hit by the previous iteration.
+        key_regexp = r'-----BEGIN ([A-Z]*).*?-----END ([A-Z].*)-----'
+        self.do_cmd_output_sub('describe', key_regexp,
+                               r'********', dotall=True)
```

Comment:
OK. So this explains why you want multiline regexes and `dotall`. I don't think it's the right way to do it however. You're kinda stretching the expected limits of the regex substitution mechanism here and we're not even really matching a regex in the strictest sense - you want to delete blocks that begin with a line starting '-----BEGIN ' and ending '-----'. That only needs fixed string matching (which is also much cheaper). The substitution is also quite unsatisfactory since it replaces a whole block of lines with a  single `********`.

I think we need a new interface to do this properly. Something like (untested / top-of-head):

```python
    def do_cmd_output_delete(cmd, start_line, end_line, replace)
```

This will find the first line matching `start_line`, and substitute from there to `end_line` with the string in `replace`. If we made it so that `start_line` replaces the 1st line and `end_line` the last we'd be able to do things like:


```python
    self.do_cmd_output_delete('describe', '-----BEGIN ', '-----END', "X" * 78)
```

Producing output like:

```
-----BEGIN 
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
-----END
```

This at least leaves some recognisable trace left in the command output - we go to some length with password and other subs to try to leave as much context for the reader as possible. The way this is done currently simply nukes the entire passage.